src/transformers/models/bart/modeling_bart.py

```diff
@@ -61,8 +62,473 @@
+        past_key_value: Optional[Tuple[torch.Tensor]] = None,
+        attn_mask: Optional[torch.Tensor] = None,
+        output_attentions: bool = False,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+        """Input shape: Batch x Time x Channel"""
+
+        # if key_value_states are provided this layer is used as a cross-attention layer
+        # for the decoder
+        is_cross_attention = key_value_states is not None
+        bsz, tgt_len, embed_dim = hidden_states.size()
+
+        # get query proj
+        query_states = self.q_proj(hidden_states) * self.scaling
+
+        # get key, value proj
+        if is_cross_attention and past_key_value is not None:
+            # reuse k,v, cross_attentions
+            key_states = past_key_value[0]
+            value_states = past_key_value[1]
+        elif is_cross_attention:
+            # cross_attentions
+            key_states = self.k_proj(key_value_states)
+            value_states = self.v_proj(key_value_states)
+        elif past_key_value is not None:
+            # reuse k, v, self_attention
+            prev_key_states = past_key_value[0]
+            prev_value_states = past_key_value[1]
+            key_states = torch.cat([prev_key_states, self.k_proj(hidden_states)], dim=1)
+            value_states = torch.cat([prev_value_states, self.v_proj(hidden_states)], dim=1)
+        else:
+            # self_attention
+            key_states = self.k_proj(hidden_states)
+            value_states = self.v_proj(hidden_states)
+
+        if self.is_decoder:
+            past_key_value = (key_states, value_states)
+
+        query_states = self._shape(query_states, tgt_len, bsz)
+        key_states = self._shape(key_states, -1, bsz)
+        value_states = self._shape(value_states, -1, bsz)
+
+        src_len = key_states.size(1)
+        attn_weights = torch.bmm(query_states, key_states.transpose(1, 2))
+
+        assert attn_weights.size() == (
+            bsz * self.num_heads,
+            tgt_len,
+            src_len,
+        ), f"Attention weights should be of size {(bsz * self.num_heads, tgt_len, src_len)}, but is {attn_weights.size()}"
+
+        if attn_mask is not None:
+            assert attn_mask.size() == (
+                bsz,
+                1,
+                tgt_len,
+                src_len,
+            ), f"Attention mask should be of size {(bsz, 1, tgt_len, src_len)}, but is {attn_mask.size()}"
+            attn_weights = attn_weights.view(bsz, self.num_heads, tgt_len, src_len) + attn_mask
+            attn_weights = attn_weights.view(bsz * self.num_heads, tgt_len, src_len)
+
+        attn_weights = F.softmax(attn_weights, dim=-1)
+
+        if output_attentions:
+            # make sure that attn_weights are included in graph
+            attn_weights_reshaped = attn_weights.view(bsz, self.num_heads, tgt_len, src_len)
+            attn_weights = attn_weights_reshaped.view(bsz * self.num_heads, tgt_len, src_len)
```

Comment:
I don't think the second line is necessary here:
```suggestion
            attn_weights_reshaped = attn_weights.view(bsz, self.num_heads, tgt_len, src_len)
```
The `attn_weights` is already of shape `(bsz * self.num_heads, tgt_len, src_len)` when entering the statement.

Reply:
yeah, that looks a bit weird. We need to do this to make sure when `output_attentions` is used that the `attn_weights` don't "loose" their gradient. This can be ensured by making sure `attn_weights_reshaped` is re-used for further computations and this is included in the graph. See `test_retain_grad_hidden_states_attentions` test.

Reply:
Oof, I see, interesting workaround. In that case, would it be possible to complete the comment
```
            # make sure that attn_weights are included in graph
```
by what you just told me?